File: algorithms/449.cpp
#include <iostream>
#include <queue>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Codec {
private:
    TreeNode *BSTinsert(TreeNode *root, int val) {
        if (root == NULL)
            root = new TreeNode(val);
        else if (val < root->val)
            root->left = BSTinsert(root->left, val);
        else
            root->right = BSTinsert(root->right, val);
        return root;
    }
public:
    // Encodes a tree to a single string.
    string serialize(TreeNode *root) {
        if (root == NULL)
            return "";
        string data = to_string(root->val);
        if (root->left)
            data = data + "," + serialize(root->left);
        if (root->right)
            data = data + "," + serialize(root->right);
        return data;
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        int n = data.size();
        if (n == 0)
            return NULL;
        TreeNode *root = NULL;
        for (int i = 0; i < n; i++) {
            int num = atoi(data.c_str()+i);
            while (i < n && data[i] != ',')
                i += 1;
            root = BSTinsert(root, num);
        }
        return root;
    }
};

int main() {
    return 0;
}

File: algorithms/python/449.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

from bisect import bisect
class Codec:
    def serialize(self, root):
        """Encodes a tree to a single string.
        :type root: TreeNode
        :rtype: str
        """
        data = []
        if not root: return ''
        def dfs(node):
            data.append(node.val)
            if node.left: dfs(node.left)
            if node.right: dfs(node.right)
        dfs(root)
        return ','.join(map(str, data))

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        :type data: str
        :rtype: TreeNode
        """
        if not data: return None
        def helper(data):
            if not data: return None
            root = TreeNode(data[0])
            bound = bisect(data, data[0], lo=1)
            root.left = helper(data[1:bound])
            root.right = helper(data[bound:])
            return root
        data = list(map(int, data.split(',')))
        return helper(data)

# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.deserialize(codec.serialize(root))
